### DISCUSSION TITLE
Programming style that should be exposed on the public API

### DISCUSSION THREAD

**Discussion by nelimee** [tqec/tqec#179] [Polls]
Following [a comment from @inmzhang](https://github.com/QCHackers/tqec/pull/173#pullrequestreview-1914500988):

> As for the plaquette hierachy, in my mind, a single Plaquette base class with some factory functions to create different library plaquettes should be enough and can eliminate the huge amount of child classes. For example:
> 
> ```py
> # only used for constructions of library plaquettes
> class PlaquetteShape(enum.Enum):
>     Square,
>     LeftRounded,
>     UpRounded,
>     RightRounded,
>     DownRounded,
> 
> make_initialization_plaquette(
>     initialization_basis: Literal['X', 'Z'],
>     plaquette_shape: PlaquetteShape, # determine the plaquette shape
> ) -> Plaquette:
>     pass
> 
> make_measurement_plaquette(
>     plaquette_shape: PlaquetteShape,
> ) -> Plaquette:
>     pass
> 
> make_syndrome_extraction_plaquette(
>     plaquette_orientation: PlaquetteShape,
>     schedule: list[int],
>     include_detector: bool = True,
>     is_first_round: bool = False,
> ) -> Plaquette:
>     pass
> ```
> The OOP-style class hierachy should work well too, I just prefer the FP way personally.

what do you prefer?

**Comment by smburdick** [comment]
As a programmer, there's something you have to consider here: does a `Plaquette` *have* a particular shape, or *is it* a particular shape?

The functional example implies the former, while the OOP approach implies the former.

Either way, it wouldn't impact the complexity much. You'd have functions that take generic `Plaquette`s and may need to check what kind of `PlaquetteShape` there for which `isinstance` is used, or which an equality comparison is. Since `isinstance` (and OOP in general in Python, imo) is so cumbersome, among other reasons (dynamic typing chief among them), many Python programmers are biased towards the functional approach, which is also inherently less bug prone, due to there less state to track.

Since correctness is really important for us in the scientific computing domain, that is why I vote for the functional approaches.

**Reply by nelimee** [reply]
> As a programmer, there's something you have to consider here: does a Plaquette have a particular shape, or is it a particular shape?

The question here is not really about shape anymore (it was before #173). Child classes of `Plaquette` are implementing a different behaviour, and are really just helpers to avoid error-prone and repetitive circuit constructions to the library end-user. 

I'll open an issue to translate the `Plaquette` hierarchy to functional-based API.

**Reply by smburdick** [reply]
> Child classes of Plaquette are implementing a different behaviour, and are really just helpers to avoid error-prone

Welcome to [the holy wars](https://dev.to/nijeesh4all/functional-programming-vs-oops--explain-x-like-im-five-2fnc#:~:text=I%20mean%20you%20can%20minimize,to%20remember%20how%20they%20work.), but I digress.

Thanks for soliciting feedback -- I enjoy these discussions, so maybe we should do more polls like this.

**Reply by nelimee** [reply]
> so maybe we should do more polls like this.

I agree. Out of curiosity, did you get a GitHub notification when the discussion/poll was open, or is it my message on the Google Group that notified you? I was under the impression that opening a discussion does not notify collaborators, and if we want to use more this feature this is the kind of things that are nice to know.

**Reply by smburdick** [reply]
I only saw it through the Google Group.

**Reply by nelimee** [reply]
Then we willneed to "advertise" discussions on the Google group, or maybe there is a setting hidden somewhere that I am not aware of.

**Reply by smburdick** [reply]
You can tag @QCHackers/tqec here as well.
